OutOfSchool/OutOfSchool.WebApi/Services/CurrentUserService.cs

```diff
@@ -0,0 +1,134 @@
+#nullable enable
+
+using System.Security.Claims;
+using OutOfSchool.Common.Models;
+using OutOfSchool.Services.Enums;
+using ProviderType = OutOfSchool.Common.Models.ProviderType;
+
+namespace OutOfSchool.WebApi.Services;
+
+public class CurrentUserService : ICurrentUserService
+{
+    private readonly ClaimsPrincipal? user;
+    private readonly IParentService parentService;
+    private readonly IProviderService providerService;
+    private readonly IProviderAdminService providerAdminService;
+
+    public CurrentUserService(
+        ClaimsPrincipal? user,
+        IProviderService providerService,
+        IProviderAdminService providerAdminService,
+        IParentService parentService)
+    {
+        this.user = user;
+        this.providerService = providerService;
+        this.providerAdminService = providerAdminService;
+        this.parentService = parentService;
+
+        // TODO: add caching as this service is going to be called everytime use makes requests
+        // TODO: so probably can cache extended user info for quick validations
+    }
+
+    public string UserId => user?.GetUserPropertyByClaimType(IdentityResourceClaimsTypes.Sub) ?? string.Empty;
+
+    public bool IsInRole(Role role) => role switch
+    {
+        Role.Provider => user?.IsInRole("provider") ?? false,
+        Role.Parent => user?.IsInRole("parent") ?? false,
+        Role.TechAdmin => user?.IsInRole("techadmin") ?? false,
+        Role.MinistryAdmin => user?.IsInRole("ministryadmin") ?? false,
+        _ => false
+    };
+
+    // TODO: in code we check multiple permissions (for example parent and provider can edit application status)
+    // TODO: either separate method calls with IsInRole check, or array like here?
+    public async Task UserHasRights(params IUserType[] userTypes)
+    {
+        // TODO: double check role?
+        var userHasRights = false;
+        if (user?.Identity?.IsAuthenticated ?? false)
+        {
+            var parent = userTypes.Select(t => t as ParentType).FirstOrDefault(p => p is not null);
+            var provider = userTypes.Select(t => t as ProviderType).FirstOrDefault(p => p is not null);
+            var providerAdmin = userTypes.Select(t => t as ProviderAdminType).FirstOrDefault(p => p is not null);
+
+            var result = await Task.WhenAll(
+                new List<Task<bool>>
+                    {
+                        UserHasRights(parent),
+                        UserHasRights(provider),
+                        UserHasRights(providerAdmin),
+                    }
+                    .Select(Execute));
+            userHasRights = result.Aggregate(false, (acc, next) => acc || next);
```

Comment:
Another code style comment, if I get it correctly all you do here is just checking for any `true` value, right? Why not then use `result.Any(hasRight => hasRight)` ?